zenoh/src/net/routing/interceptor/authorization.rs

```diff
@@ -32,7 +35,81 @@ use zenoh_util::net::get_interface_names_by_addr;
 type PolicyForSubject = FlowPolicy;
 
 type PolicyMap = HashMap<usize, PolicyForSubject, RandomState>;
-type SubjectMap = HashMap<Subject, usize, RandomState>;
+
+#[derive(Debug, Clone)]
+pub(crate) struct SubjectMap {
+    inner: TrieMap<Subject, usize>,
+}
+
+impl SubjectMap {
+    pub(crate) fn builder() -> SubjectMapBuilder {
+        SubjectMapBuilder::new()
+    }
+
+    pub(crate) fn get(&self, subjects: &[Subject]) -> Vec<(Vec<Subject>, &usize)> {
+        let mut subjects = subjects.to_owned();
+        subjects.sort_unstable();
+        let mut res: Vec<(Vec<Subject>, &usize)> = vec![];
+        for subject in subjects {
+            let map_query: Vec<(Vec<Subject>, &usize)> =
+                self.inner.predictive_search([subject]).collect();
+            for (subject, id) in map_query {
+                if subject.len() > 1 {
+                    for s in &subject {
+                        if subject.binary_search(s).is_err() {
```

Comment:
Isn't this always true?

Reply:
@fuzzypixelz I don't see why it would be always false, could you elaborate?

Reply:
Because `subject.contains(s)` as `s` is the iteration variable over `subjects`?

Reply:
Ah there's a typo, it should've been `subjects.binary_search(s).is_err()`.
`subject` is a subject combination yielded from the predictive search within the trie, it needs to be checked if it contains subjects that are not present in the query, hence the search in `subjects` list.
The variable names are a bit confusing...

Reply:
Okay, this is likely the source of the bug I observed. I honestly couldn't be sure of the intent behind the code, so it was simpler for me to replace it for now. It's ultimately up to you to decide on an (efficient) data structure.

Reply:
I'd propose to focus on the correctness of this PR at this stage.
Configuration parsing could be further optimized but it should produce correct result first. 
Therefore I suggest to go ahead with the solution proposed by @fuzzypixelz if proved correctly at this stage.